# Assignment 2.

## Formalia:

Please read the [assignment overview page](https://github.com/suneman/socialdataanalysis2020/wiki/Assignments) carefully before proceeding. This page contains information about formatting (including formats etc), group sizes, and many other aspects of handing in the assignment. 

_If you fail to follow these simple instructions, it will negatively impact your grade!_

**Due date and time**: The assignment is due on Monday March 30th, 2020 at 23:55. Hand in your files via [`http://peergrade.io`](http://peergrade.io/).

**Peergrading date and time**: _Remember that after handing in you have 6 days to evaluate a few assignments written by other members of the class_. Thus, the peer evaluations are due on Sunday April 5th, 2020 at 23:55. (The reason I'm asking you to do the peer evaluations on the sunday is that I want you to be able to enjoy the easter break fully.)

In [2]:
import sys 
from termcolor import colored, cprint 
from PIL import Image
import pandas as pd
import numpy as np
import datetime as dt
from bokeh.plotting import figure,show
from bokeh.io import output_notebook
from bokeh.models import ColumnDataSource, FactorRange,Legend
from bokeh.palettes import Category20
from bokeh.plotting import figure
from bokeh.transform import factor_cmap
output_notebook()

Loading BokehJS ...

## Part 1: Questions to text and lectures.

**A) Please answer my questions to the Segal and Heer paper we read during lecture 7 and 8.**

* What is the Oxford English Dictionary's defintion of a narrative?

<font color='green'>A spoken or written account of connected events, a story.</font>

* What is your favorite visualization among the examples in section 3? Explain why in a few words.

<font color='green'>Our favorite is *Fig. 3. Afghanistan: Behind the Front Line. Financial Times*, because is has a clear, easy and consistent visualization supported by the color indication.
    
It is clear that clicking on the individual country highlights the provinces in which that
country has troops deployed, outlining their borders with a red line. 
Furthermore, the tabs makes it possible to explore individual contries.


However, it might be hard to draw meaningful conclusions from the large amount of information available.</font>

* What's the point of Figure 7?

<font color='green'>The point of Figure 7 is to represent the specific design space strategies observed across dataset.
<br>
    
The design space contains three divisions of features: 
    <br>
    (1) genre,
    <br>
    (2) visual narrative tactics, and
    <br>
    (3) narrative structure tactics. 
    <br>
<br>
The first division identifies the genre of each visualization like Magazine Style, Flow Chart, Slide Show etc.
<br>
The second division identifies visual narrative tactics: visual devices that assist and facilitate the narrative. This division is subdivided into three sections: 
<br>
    (i) visual structuring, 
<br>
    (ii) highlighting, and
<br>
    (iii) transition guidance. 
<br>
<br>
**Visual structuring** refers to mechanisms that communicate the overall structure of the narrative to the viewer. 
<br>
**Highlighting** refers to visual mechanisms that help direct the viewer’s attention to particular elements in the display.
<br>
**Transition guidance** concerns techniques for moving within or between visual scenes without disorienting the viewer.
<br>
<br>
The third division identifies narrative structure tactics used by each visualization, or non-visual mechanisms that assist and facilitate the narrative. This division is further divided into three sections: 
<br>
(i) ordering, 
<br>
(ii) interactivity, and 
<br>
(iii) messaging. 
<br>

**Ordering** refers to the ways  of arranging the path viewers take through the visualization. 
<br>
**Interactivity** refers to the different ways a user can manipulate the visualization, and also how the user learns those methods.
<br>
**Messaging** refers to the ways a visualization communicates observations and commentary to the viewer.

In Figure 7 is the specific design space strategies are observed across dataset and categories in each field.
</font>

* Use Figure 7 to find the most common design choice within each category for the Visual narrative and Narrative structure (the categories within visual narrative are 'visual structuring', 'highlighting', etc).


<font color='green'> 
The common design choices in each category is marked with red text:    
<br>
Visual narrative tactics:
<br>
(i) visual structuring; <font color='red'>Consistent Visual Platform </font>
<br>
(ii) highlighting;  <font color='red'>Feature Distinction</font>
<br>
(iii) transition guidance; <font color='red'>Animated Transitions</font>
<br>
Narrative structure tactics: 
<br>
(i) ordering; <font color='red'>Filtering / Selection / Search</font>
<br>
(ii) interactivity; <font color='red'>Filtering / Selection / Search</font>
<br>
(iii) messaging. <font color='red'>Captions / Headlines</font>
<br>

</font>

* Check out Figure 8 and section 4.3. What is your favorite genre of narrative visualization? Why? What is your least favorite genre? Why?


<font color='green'> It is diffucult to only choose one favorite genre of narrative visualization, and as it stated in the text the genres are not mutually exclusive and they can function like
building blocks. 
<br>
**Our favorite is** <font color='red'>Annotated Graph, Flow Charts and Slide Show.</font> 
<br>
The reason for that is; these can be more informative and the message is better interpreted.   
<br>
<br>
**Our least favorite is** <font color='red'>Magazine Style.</font> 
<br>
The reason for that is; the viewer will start reading all the text which might be summarized in some figures and less text. This will increase the time for the viewer to understand, but on the other hand makes sure that it's fully understood.

<br>
<br>
Bear in mind, that some genres can be more useful depending on the message and the target group. 

</font>

**B) Also please answer the questions to my talk on [explanatory data visualization](https://www.youtube.com/watch?v=yHKYMGwefso)**


* What are the three key elements to keep in mind when you design an explanatory visualization?


<font color='green'> 
* Start with a question (what is the result you want to communicate?)
* Allow exploration (let the users investigate...)
* Know your readers (design for an audience). 

</font>

* In the video I talk about (1) *overview first*,  (2) *zoom and filter*,  (3) *details on demand*. 
  - Go online and find a visualization that follows these principles (don't use one from the video). 
  - Explain how it does achieves (1)-(3). It might be useful to use screenshots to illustrate your explanation.



<font color='green'> To answer the above questions, following website has been used: https://coinmarketcap.com/currencies/bitcoin/

<br>
<br>
When working with stocks (in our case), it is important to get the overview for the prefered stock, in our case Bitcoin:
<br>

- The shown with a graph, which gives a historically overview on the development of the stock. <font color='red'> (*(1) overview first*)</font>
<br>

- Furthermore, the graph allows zoom and filtering, so you can choose to see different time periods; 1 day, 7 days, 1 month, 3 month, 1 year, year to date and finally all data. <font color='red'> (*(2) zoom and filter*)</font>
<br>

- Finally, it is possible to get further details of the data by clicking on the tabs above the graph <font color='red'> (*(3) details on demand* )</font>
</font>

</font>

<img src="https://i.ibb.co/4gCx7s5/bitcoin.png" alt="bitcoin" width="650"/>

* Explain in your own words: How is explanatory data analysis different from exploratory data analysis?

<font color='green'> **Exploratory data analysis**, is to explore the data to become familiar with it, and focuses on hypothesis formulation and digging through your data (data analysis) to find relationship and trends without a specific goal.
    
**Explanatory visualization** focuses on the communication results, meaning that you have to explain outcomes or relationships of something specific you want to show an audiance / targetgroup by telling a story with your data. 
</font>

## Part 2: Random forest and weather

The aim here is to recreate the work you did in Part 1 and 2 of the Week 7 lecture. I've phrased things differently relative to the exercise to make the purpose more clear. 

Part 2A: Random forest binary classification. 
* Using the and instructions and material from Week 7, build a *random forest* classifier to distinguish between two types (you choose) of crime using on spatio-temporal (where/when) features of data describing the two crimes. When you're done, you should be able to give the classifier a place and a time, and it should tell you which of the two  types of crime happened there.
  - Explain about your choices for training/test data, features, and encoding. (You decide how to present your results, but here are some example topics to consider: Did you balance the training data? What are the pros/cons of balancing? Do you think your model is overfitting? Did you choose to do cross-validation? Which specific features did you end up using? Why? Which features (if any) did you one-hot encode? Why ... or why not?))
  - Report accuracy. Discuss the model performance.
  
  


In [3]:
# we start by importing the packages that will be used
import calendar
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import RandomizedSearchCV

In [4]:
df= pd.read_csv('Police_Department_Incident_Reports__Historical_2003_to_May_2018.csv')

<font color='green'> We will work with **BURGLARY** and **FRAUD** due to the fact that there are many crimes of this types.
</font>

In [5]:
crimes=set(['BURGLARY', 'FRAUD'])

# we get the 168 hours of the week
df['Time'] = [x + ':00' for x in df['Time']]
df['TimeInt']=pd.to_timedelta(df.Time, unit='hours').dt.floor('H')
df['HourInt']=  df['TimeInt']/ np.timedelta64(1, 'h')

days = dict(zip(calendar.day_name, range(7))); 
df['Day_nr'] = df['DayOfWeek'].map(lambda x: days[x])
# hour 0 corresponds to monday 00:00 - 01:00
df['Interval168']=df['Day_nr']*24+df['HourInt']

# df_rf corresponds to the dataframe that will be used on the random forest classifier
df_rf=df[df.Category.isin(crimes)]

In [6]:
# we can check how many of each category of crime do we have and also drop NaNs
df_rf=df_rf.dropna()
df_rf['Category'].value_counts()

BURGLARY    91543
FRAUD       41542
Name: Category, dtype: int64

<font color='green'> Now, it is time to encode the attribute `PdDistrict`, first we will use a scikit's LabelEncoder, which will transform the `string` into an integer, so that we can pass it to our random forest classifier.
    
Later on we will do the transformation using one-hot encoding and we'll compare the performance of our classifier. What do you think will happen? Let's find out 😀
</font>

In [7]:
le = preprocessing.LabelEncoder()
le.fit(df_rf['PdDistrict'])
df_rf['PdDistrict_nr'] = le.transform(df_rf['PdDistrict'])

In [8]:
'''Here we select the features that we want to use on the classifier
which are the 168 hours of the week and the district.'''
df_X1=df_rf[['Interval168','PdDistrict_nr']]
df_y1=df_rf['Category']

#then we transform it into a numpy array, since that is what the classifier "receives"
X1=df_X1.to_numpy()
y1=df_y1.to_numpy()

''' Separation of the dataset between training and test data.
the test size is 33% of the whole dataset.'''
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.33, random_state=42)

unique_elements, counts_elements = np.unique(y1_train, return_counts=True)

print("Frequency of values on the train dataset")
print(np.asarray((unique_elements, counts_elements)))

#we build the random forest
rf1 = RandomForestClassifier(n_estimators=100,random_state = 42,class_weight='balanced')

rf1.fit(X1_train,y1_train)
y1_pred=rf1.predict(X1_test)

print('Accuracy score:')
print(accuracy_score(y1_test, y1_pred))
print('Classification report on test data:')
print(classification_report(y1_test, y1_pred))
print('Confusion Matrix:')
print(confusion_matrix(y1_test, y1_pred))
print('Classification report on train data (check for overfitting):')
print(classification_report(rf1.predict(X1_train), y1_train))


# we can verify the precision with a function defined by us as well
def myprecision(y,ypred, classlist):
    TP,FP=[0,0],[0,0]
    for i in range(len(y)):
        if ypred[i]==y[i]:
            if ypred[i]==classlist[0]:
                TP[0]+=1
            else:
                TP[1]+=1
        else:
            if ypred[i]==classlist[0]:
                FP[0]+=1
            else:
                FP[1]+=1
    return [classlist[i]+': '+str(TP[i]/(FP[i]+TP[i])) for i in range(len(TP))]

myprecision(y1_test,y1_pred,['BURGLARY','FRAUD'])

Frequency of values on the train dataset
[['BURGLARY' 'FRAUD']
 [61317 27849]]
Accuracy score:
0.6415446617637014
Classification report on test data:
              precision    recall  f1-score   support

    BURGLARY       0.79      0.66      0.72     30226
       FRAUD       0.45      0.61      0.51     13693

    accuracy                           0.64     43919
   macro avg       0.62      0.63      0.62     43919
weighted avg       0.68      0.64      0.65     43919

Confusion Matrix:
[[19837 10389]
 [ 5354  8339]]
Classification report on train data (check for overfitting):
              precision    recall  f1-score   support

    BURGLARY       0.67      0.80      0.73     51064
       FRAUD       0.63      0.46      0.54     38102

    accuracy                           0.66     89166
   macro avg       0.65      0.63      0.63     89166
weighted avg       0.65      0.66      0.65     89166



['BURGLARY: 0.7874637767456631', 'FRAUD: 0.44526911576249467']

<font color='green'> So it is time to check how the model performs with one-hot encoding by using pandas get_dummies function.
</font>

In [9]:
df_rf = pd.concat([df_rf,pd.get_dummies(df_rf['PdDistrict'], prefix='PdDistrict')],axis=1)
list(df_rf.columns)

['IncidntNum',
 'Category',
 'Descript',
 'DayOfWeek',
 'Date',
 'Time',
 'PdDistrict',
 'Resolution',
 'Address',
 'X',
 'Y',
 'Location',
 'PdId',
 'TimeInt',
 'HourInt',
 'Day_nr',
 'Interval168',
 'PdDistrict_nr',
 'PdDistrict_BAYVIEW',
 'PdDistrict_CENTRAL',
 'PdDistrict_INGLESIDE',
 'PdDistrict_MISSION',
 'PdDistrict_NORTHERN',
 'PdDistrict_PARK',
 'PdDistrict_RICHMOND',
 'PdDistrict_SOUTHERN',
 'PdDistrict_TARAVAL',
 'PdDistrict_TENDERLOIN']

In [10]:
'''Here we select the features that we want to use on the classifier
which are the 168 hours of the week and the district.'''
df_X2=df_rf[['Interval168', 'PdDistrict_BAYVIEW','PdDistrict_CENTRAL','PdDistrict_INGLESIDE','PdDistrict_MISSION','PdDistrict_NORTHERN',
             'PdDistrict_PARK','PdDistrict_RICHMOND','PdDistrict_SOUTHERN','PdDistrict_TARAVAL','PdDistrict_TENDERLOIN']]
df_y2=df_rf['Category']


X2=df_X2.to_numpy()
y2=df_y2.to_numpy()


X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.33, random_state=42)

#we build the random forest
rf2 = RandomForestClassifier(n_estimators=100,random_state = 42,class_weight='balanced')

rf2.fit(X2_train,y2_train)
y2_pred=rf2.predict(X2_test)

print(accuracy_score(y2_test, y2_pred))
print(classification_report(y2_test, y2_pred))
print(confusion_matrix(y2_test, y2_pred))

0.6415446617637014
              precision    recall  f1-score   support

    BURGLARY       0.79      0.66      0.72     30226
       FRAUD       0.45      0.61      0.51     13693

    accuracy                           0.64     43919
   macro avg       0.62      0.63      0.62     43919
weighted avg       0.68      0.64      0.65     43919

[[19837 10389]
 [ 5354  8339]]


<font color='green'> So, the random forest classifier give us **exactly** the same results when using LabelEncoder and one-hot encoding. The explanation lies in that RandomForestClassifier does not handles the values on PdDistrict_nr in the first classifier to be **x** times as good a as another one (if PdDistrict_nr is 1 or 9 it does not affect the classification). It just finds the best splits to minimize entropy.
    
Now, we are going to use CV to determine the best model features.
</font>

In [11]:
from pprint import pprint
# Look at parameters used by our current forest
print('Parameters currently in use:\n')
pprint(rf1.get_params())

Parameters currently in use:

{'bootstrap': True,
 'class_weight': 'balanced',
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': 42,
 'verbose': 0,
 'warm_start': False}


In [12]:
# Number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 10, stop = 500, num = 50)]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 80, num = 5)]
max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4]
# Method of selecting samples for training each tree
bootstrap = [True, False]
# Create the random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}
pprint(random_grid)

{'bootstrap': [True, False],
 'max_depth': [10, 27, 45, 62, 80, None],
 'max_features': ['auto', 'sqrt'],
 'min_samples_leaf': [1, 2, 4],
 'min_samples_split': [2, 5, 10],
 'n_estimators': [10,
                  20,
                  30,
                  40,
                  50,
                  60,
                  70,
                  80,
                  90,
                  100,
                  110,
                  120,
                  130,
                  140,
                  150,
                  160,
                  170,
                  180,
                  190,
                  200,
                  210,
                  220,
                  230,
                  240,
                  250,
                  260,
                  270,
                  280,
                  290,
                  300,
                  310,
                  320,
                  330,
                  340,
                  350,
                  360,
         

In [13]:
'''Careful here, this takes around 8 minutes to run,
so either trust our output or grab a coffee and wait for the results :)'''
# Use the random grid to search for best hyperparameters
# First create the base model to tune
rfcv = RandomForestClassifier(class_weight='balanced')
# Random search of parameters, using 3 fold cross validation, 
# search across 100 different combinations, and use all available cores
rf_random = RandomizedSearchCV(estimator = rfcv, param_distributions = random_grid, n_iter = 50, cv = 3, verbose=2, random_state=42, n_jobs = -1)
# Fit the random search model
rf_random.fit(X1_train, y1_train)

Fitting 3 folds for each of 50 candidates, totalling 150 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 16 concurrent workers.
[Parallel(n_jobs=-1)]: Done   9 tasks      | elapsed:   27.6s
[Parallel(n_jobs=-1)]: Done 150 out of 150 | elapsed:  4.6min finished


RandomizedSearchCV(cv=3, error_score='raise-deprecating',
                   estimator=RandomForestClassifier(bootstrap=True,
                                                    class_weight='balanced',
                                                    criterion='gini',
                                                    max_depth=None,
                                                    max_features='auto',
                                                    max_leaf_nodes=None,
                                                    min_impurity_decrease=0.0,
                                                    min_impurity_split=None,
                                                    min_samples_leaf=1,
                                                    min_samples_split=2,
                                                    min_weight_fraction_leaf=0.0,
                                                    n_estimators='warn',
                                                    n_job

In [14]:
rf_random.best_params_

{'n_estimators': 440,
 'min_samples_split': 10,
 'min_samples_leaf': 4,
 'max_features': 'sqrt',
 'max_depth': None,
 'bootstrap': True}

In [15]:
best_random = rf_random.best_estimator_

y_pred_cv=best_random.predict(X1_test)


print(accuracy_score(y1_test, y_pred_cv))
print(classification_report(y1_test, y_pred_cv))
print(confusion_matrix(y1_test, y_pred_cv))
myprecision(y1_test,y_pred_cv,['BURGLARY','FRAUD'])


0.6430929665976002
              precision    recall  f1-score   support

    BURGLARY       0.79      0.66      0.72     30226
       FRAUD       0.45      0.61      0.51     13693

    accuracy                           0.64     43919
   macro avg       0.62      0.63      0.62     43919
weighted avg       0.68      0.64      0.65     43919

[[19946 10280]
 [ 5395  8298]]


['BURGLARY: 0.7871039027662681', 'FRAUD: 0.44665733663472923']

<font color='green'> We can see that tuning the parameters of the model by using cross validation improves the performance of the model against previously unseen data (test set). But, the improvement not a big one. we can see that the precision changes from:['BURGLARY: 0.7874637767456631', 'FRAUD: 0.44526911576249467'] to ['BURGLARY: 0.7878907639410346', 'FRAUD: 0.44724967769660506'] which is an improvement of 0.054% and 0.44% of the prediction of BURGLARY and FRAUD, respectively.


We kept this model simple as we want to see if the weather is actually correlated with the occurrence of crimes of type 1. We chose to balance the training data using `class_weight='balanced'`,  the reason is that there is almost double of frequency of BURGLARY compared to FRAUD, meaning that the classifier would lean towards classifyng according to the most repeated category, rather than centering in the features to classify the data. By doing this, it is expected that the misclassifications are somewhat similar between the two type of crimes.

It appears that the model does not overfit, we can check it by comparinf the classification report  on the train set and the test set. The performance of the model is not considerable different between them, which suggest that the model is not overfitting and learning the data instead of learnin _from_ the data.

As we explained before, we do get better results when using cross-validation at expense of the model taking more time to run. 

We explained and show how one-hot encoding did not affect the performance of the model. In case you skipped the reading earlier in the notebook :) . 

</font>

Part 2B: Info from weather features.
* Now add features from weather data to your random forest. 
  - Report accuracy. 
  - Discuss how the model performance changes relative to the version with no weather data.
  - Discuss what you have learned about crime from including weather data in your model.

In [16]:
df_rf2=df[df.Category.isin(crimes)]
df_rf2=df_rf2.dropna()
df_rf2['DateTime']= pd.to_datetime(df_rf2['Date']+' '+df_rf2['Time'])
df_rf2.head(2)

IncidntNum  Category                                     Descript  \
11   110318854     FRAUD                FRAUDULENT CREDIT APPLICATION   
32   150101346  BURGLARY  BURGLARY OF APARTMENT HOUSE, UNLAWFUL ENTRY   

   DayOfWeek        Date      Time PdDistrict Resolution  \
11  Thursday  04/14/2011  17:30:00    BAYVIEW       NONE   
32    Monday  02/02/2015  16:30:00    TARAVAL       NONE   

                   Address           X          Y  \
11  100 Block of TOPEKA AV -122.398951  37.733847   
32   2400 Block of 41ST AV -122.499265  37.741054   

                                      Location            PdId  TimeInt  \
11  POINT (-122.398951090192 37.7338473162583)  11031885409340 17:00:00   
32  POINT (-122.499264595602 37.7410543982056)  15010134605013 16:00:00   

    HourInt  Day_nr  Interval168            DateTime  
11     17.0       3         89.0 2011-04-14 17:30:00  
32     16.0       0         16.0 2015-02-02 16:30:00

In [17]:
import urllib.request as urllib2

url = 'https://raw.githubusercontent.com/suneman/socialdataanalysis2020/master/files/Data_files/weather_data.csv'
response = urllib2.urlopen(url)
weather = pd.read_csv(response)

weather['DateTime'] = pd.to_datetime(weather['date'],format='%Y-%m-%dT%H:%M:%S.%fZ')
weather = weather.rename(columns={'date':'Date'})
weather = pd.get_dummies(weather, columns=["weather"], prefix=["w"])

len(weather) # = 44306
#Rounding hour to nearest hour in data_tree
df_rf2.DateTime = df_rf2.DateTime.dt.round('H')


df_rf2.sample(1)

IncidntNum  Category                           Descript DayOfWeek  \
1092554    60335126  BURGLARY  BURGLARY OF STORE, FORCIBLE ENTRY   Tuesday   

               Date      Time PdDistrict Resolution              Address  \
1092554  03/28/2006  18:00:00   RICHMOND       NONE  47TH AV / FULTON ST   

                 X          Y                                    Location  \
1092554 -122.50775  37.771494  POINT (-122.507750131004 37.7714939969416)   

                  PdId  TimeInt  HourInt  Day_nr  Interval168  \
1092554  6033512605051 18:00:00     18.0       1         42.0   

                   DateTime  
1092554 2006-03-28 18:00:00

In [18]:
weather.sample(1)

Date  temperature  humidity  wind_speed  \
24267  2015-07-16T01:00:00.000Z        26.32      89.0         2.0   

       wind_direction  pressure            DateTime  w_broken clouds  \
24267           340.0    1011.0 2015-07-16 01:00:00                0   

       w_drizzle  w_few clouds  ...  w_scattered clouds  w_shower rain  \
24267          0             0  ...                   0              0   

       w_sky is clear  w_smoke  w_squalls  w_thunderstorm  \
24267               0        0          0               0   

       w_thunderstorm with heavy rain  w_thunderstorm with light rain  \
24267                               0                               0   

       w_thunderstorm with rain  w_very heavy rain  
24267                         0                  0  

[1 rows x 35 columns]

In [19]:
#merge weather df with crime data
df_rf2w = pd.merge(df_rf2,weather,how='inner',on='DateTime')
df_rf2w.Category.value_counts()

BURGLARY    30366
FRAUD       14384
Name: Category, dtype: int64

In [20]:
df_rf2w = pd.concat([df_rf2w,pd.get_dummies(df_rf2w['PdDistrict'], prefix='PdDistrict')],axis=1)
print(list(df_rf2w.columns))

['IncidntNum', 'Category', 'Descript', 'DayOfWeek', 'Date_x', 'Time', 'PdDistrict', 'Resolution', 'Address', 'X', 'Y', 'Location', 'PdId', 'TimeInt', 'HourInt', 'Day_nr', 'Interval168', 'DateTime', 'Date_y', 'temperature', 'humidity', 'wind_speed', 'wind_direction', 'pressure', 'w_broken clouds', 'w_drizzle', 'w_few clouds', 'w_fog', 'w_haze', 'w_heavy intensity drizzle', 'w_heavy intensity rain', 'w_heavy snow', 'w_light intensity drizzle', 'w_light intensity shower rain', 'w_light rain', 'w_light snow', 'w_mist', 'w_moderate rain', 'w_overcast clouds', 'w_proximity shower rain', 'w_proximity thunderstorm', 'w_proximity thunderstorm with rain', 'w_scattered clouds', 'w_shower rain', 'w_sky is clear', 'w_smoke', 'w_squalls', 'w_thunderstorm', 'w_thunderstorm with heavy rain', 'w_thunderstorm with light rain', 'w_thunderstorm with rain', 'w_very heavy rain', 'PdDistrict_BAYVIEW', 'PdDistrict_CENTRAL', 'PdDistrict_INGLESIDE', 'PdDistrict_MISSION', 'PdDistrict_NORTHERN', 'PdDistrict_PARK'

In [21]:
df_rf2w.dropna()
df_rf2w_X=df_rf2w[['Interval168', 'w_broken clouds','w_drizzle','w_few clouds','w_fog','w_haze','w_heavy intensity drizzle',
'w_heavy intensity rain','w_heavy snow','w_light intensity drizzle','w_light intensity shower rain','w_light rain','w_light snow','w_mist',
'w_moderate rain','w_overcast clouds','w_proximity shower rain','w_proximity thunderstorm','w_proximity thunderstorm with rain','w_scattered clouds','w_shower rain','w_sky is clear','w_smoke','w_squalls',
 'w_thunderstorm','w_thunderstorm with heavy rain','w_thunderstorm with light rain','w_thunderstorm with rain','w_very heavy rain','PdDistrict_BAYVIEW',
 'PdDistrict_CENTRAL','PdDistrict_INGLESIDE','PdDistrict_MISSION','PdDistrict_NORTHERN','PdDistrict_PARK','PdDistrict_RICHMOND','PdDistrict_SOUTHERN','PdDistrict_TARAVAL','PdDistrict_TENDERLOIN',
                   'temperature','humidity','wind_speed','wind_direction','pressure']]
df_rf2w_X.sample(5)

Interval168  w_broken clouds  w_drizzle  w_few clouds  w_fog  w_haze  \
11265        147.0                0          0             0      0       0   
34365         66.0                0          0             1      0       0   
31613        114.0                0          0             0      0       0   
24370         56.0                1          0             0      0       0   
15245         55.0                0          0             0      0       0   

       w_heavy intensity drizzle  w_heavy intensity rain  w_heavy snow  \
11265                          0                       0             0   
34365                          0                       0             0   
31613                          0                       0             0   
24370                          0                       0             0   
15245                          0                       0             0   

       w_light intensity drizzle  ...  PdDistrict_PARK  PdDistrict_RICHMOND  \
11265                          0  ...                0                    0   
34365                          0  ...                0                    0   
31613                          0  ...                0                    0   
24370                          0  ...                0                    0   
15245                          0  ...                0                    0   

       PdDistrict_SOUTHERN  PdDistrict_TARAVAL  PdDistrict_TENDERLOIN  \
11265                    0                   0                      0   
34365                    0                   0                      0   
31613                    0                   0                      0   
24370                    1                   0                      0   
15245                    0                   0                      0   

       temperature  humidity  wind_speed  wind_direction  pressure  
11265    13.604333      84.0         2.0           293.0    1030.0  
34365    20.820000      62.0         6.0           120.0    1017.0  
31613    19.280000      77.0         3.0           300.0    1014.0  
24370    13.770000      89.0         1.0           232.0    1015.0  
15245    11.231667      97.0         2.0           225.0    1023.0  

[5 rows x 44 columns]

In [22]:

df_rf2w_y=df_rf2w['Category']

def getScores(dfx,dfy):
    X=dfx.to_numpy()
    y=dfy.to_numpy()
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
    
    nanIndex = np.argwhere(np.isnan(X_train))
    nanIndex = np.delete(nanIndex, 1, 1)
    nanIndex = np.unique(nanIndex)
    X_train = np.delete(X_train, nanIndex, axis=0)
    y_train =  np.delete(y_train, nanIndex)
    rff = RandomForestClassifier(n_estimators=100,random_state = 42, class_weight='balanced')
    rff.fit(X_train,y_train)
    y_pred=rff.predict(X_test)
    print('Accuracy score:')
    print(accuracy_score(y_test, y_pred))
    print('Classification report on test data:')
    print(classification_report(y_test, y_pred))
    print('Confusion Matrix:')
    print(confusion_matrix(y_test, y_pred))
    print('Classification report on train data (check for overfitting):')
    print(classification_report(rff.predict(X_train), y_train))

getScores(df_rf2w_X,df_rf2w_y)

Accuracy score:
0.6771397616468039
Classification report on test data:
              precision    recall  f1-score   support

    BURGLARY       0.71      0.90      0.79     10088
       FRAUD       0.48      0.20      0.28      4680

    accuracy                           0.68     14768
   macro avg       0.59      0.55      0.54     14768
weighted avg       0.63      0.68      0.63     14768

Confusion Matrix:
[[9059 1029]
 [3739  941]]
Classification report on train data (check for overfitting):
              precision    recall  f1-score   support

    BURGLARY       0.98      0.99      0.99     19940
       FRAUD       0.99      0.96      0.97     10041

    accuracy                           0.98     29981
   macro avg       0.98      0.98      0.98     29981
weighted avg       0.98      0.98      0.98     29981



<font color='green'> The average accuracy actually improved by adding the weather data to the predictions. But all the other measures got actually worse (precision, recall and f1-score).

Now, let's see how if by enriching the data just with the descriptive weather actually improves the performance of the model. Then, we will just check with only three parameters, by adding just the temperature to the data and see if we get a better prediction.

</font>

In [23]:
df_rf2w_X1=df_rf2w[['Interval168','PdDistrict_BAYVIEW',
 'PdDistrict_CENTRAL','PdDistrict_INGLESIDE','PdDistrict_MISSION','PdDistrict_NORTHERN','PdDistrict_PARK',
                    'PdDistrict_RICHMOND','PdDistrict_SOUTHERN','PdDistrict_TARAVAL','PdDistrict_TENDERLOIN',
                   'w_broken clouds','w_drizzle','w_few clouds','w_fog','w_haze','w_heavy intensity drizzle',
'w_heavy intensity rain','w_heavy snow','w_light intensity drizzle','w_light intensity shower rain','w_light rain','w_light snow','w_mist',
'w_moderate rain','w_overcast clouds','w_proximity shower rain','w_proximity thunderstorm','w_proximity thunderstorm with rain','w_scattered clouds','w_shower rain','w_sky is clear','w_smoke','w_squalls',
 'w_thunderstorm','w_thunderstorm with heavy rain','w_thunderstorm with light rain','w_thunderstorm with rain','w_very heavy rain']]


df_rf2w_y1=df_rf2w['Category']
getScores(df_rf2w_X1,df_rf2w_y1)

Accuracy score:
0.5979821235102926
Classification report on test data:
              precision    recall  f1-score   support

    BURGLARY       0.73      0.65      0.69     10088
       FRAUD       0.39      0.49      0.43      4680

    accuracy                           0.60     14768
   macro avg       0.56      0.57      0.56     14768
weighted avg       0.62      0.60      0.61     14768

Confusion Matrix:
[[6549 3539]
 [2398 2282]]
Classification report on train data (check for overfitting):
              precision    recall  f1-score   support

    BURGLARY       0.79      0.88      0.84     18157
       FRAUD       0.78      0.64      0.71     11825

    accuracy                           0.79     29982
   macro avg       0.79      0.76      0.77     29982
weighted avg       0.79      0.79      0.79     29982



In [24]:
df_rf2w_X2=df_rf2w[['Interval168','PdDistrict_BAYVIEW','PdDistrict_CENTRAL','PdDistrict_INGLESIDE',
                    'PdDistrict_MISSION','PdDistrict_NORTHERN','PdDistrict_PARK',
                    'PdDistrict_RICHMOND','PdDistrict_SOUTHERN','PdDistrict_TARAVAL','PdDistrict_TENDERLOIN',
                    'temperature'
                   ]]

getScores(df_rf2w_X2,df_rf2w_y1)


Accuracy score:
0.6409804983748646
Classification report on test data:
              precision    recall  f1-score   support

    BURGLARY       0.72      0.78      0.75     10088
       FRAUD       0.42      0.35      0.38      4680

    accuracy                           0.64     14768
   macro avg       0.57      0.56      0.56     14768
weighted avg       0.62      0.64      0.63     14768

Confusion Matrix:
[[7849 2239]
 [3063 1617]]
Classification report on train data (check for overfitting):
              precision    recall  f1-score   support

    BURGLARY       0.98      0.99      0.99     19974
       FRAUD       0.99      0.96      0.97     10008

    accuracy                           0.98     29982
   macro avg       0.98      0.97      0.98     29982
weighted avg       0.98      0.98      0.98     29982



<font color='green'> It is not a big change to add the weather data to the dataset, we have two theories on why this is happening:
    1. The crimes do not correlate with the weather.
    2. Since the weather data was only for a couple of years, the datapoints got reduced signicantly, having approximately 33% of the data compared with part one of this section.

In order to discard or approve hypothesis 2, we are going to check the prediction for the same data entries that we use in this section, but now, we only consider the hour and the district.

</font>

In [25]:
df_rf2w_Xf=df_rf2w[['Interval168','PdDistrict_BAYVIEW','PdDistrict_CENTRAL','PdDistrict_INGLESIDE',
                    'PdDistrict_MISSION','PdDistrict_NORTHERN','PdDistrict_PARK',
                    'PdDistrict_RICHMOND','PdDistrict_SOUTHERN','PdDistrict_TARAVAL','PdDistrict_TENDERLOIN'
                   ]]

getScores(df_rf2w_Xf,df_rf2w_y1)

Accuracy score:
0.6138271939328277
Classification report on test data:
              precision    recall  f1-score   support

    BURGLARY       0.78      0.61      0.68     10088
       FRAUD       0.43      0.62      0.51      4680

    accuracy                           0.61     14768
   macro avg       0.60      0.62      0.59     14768
weighted avg       0.67      0.61      0.63     14768

Confusion Matrix:
[[6142 3946]
 [1757 2923]]
Classification report on train data (check for overfitting):
              precision    recall  f1-score   support

    BURGLARY       0.65      0.81      0.72     16246
       FRAUD       0.68      0.48      0.57     13736

    accuracy                           0.66     29982
   macro avg       0.67      0.65      0.64     29982
weighted avg       0.67      0.66      0.65     29982



<font color='green'> Now it is quite clear that hypothesis 2 is the correct one and that the crime data effectively correlates with the weather. The accuracy of the model improves substantially when we add the weather data and it goes from 61% to 68%.
    It is important to note that having larger datasets makes your model richer and that was what is proven in this exercise.

</font>

## Part 3: Data visualization

* Create the Bokeh visualization from Part 2 of the Week 8 Lecture, displayed in a beautiful `.gif` below. 
* Provide nice comments for your code. Don't just use the `# inline comments`, but the full Notebook markdown capabilities and explain what you're doing.

![Movie](https://github.com/suneman/socialdataanalysis2020/blob/master/files/week8_1.gif?raw=true "movie")

## Creating the same plot as above with comments


#### 1. Data is fetched and proccessed
Fetching the crime data and previewing the first lines of the dataframe

In [26]:
crime = pd.read_csv("Police_Department_Incident_Reports__Historical_2003_to_May_2018.csv", sep=",")
crime.head()

IncidntNum        Category                 Descript  DayOfWeek        Date  \
0   146196161    NON-CRIMINAL            LOST PROPERTY    Tuesday  09/23/2014   
1   150045675         ASSAULT                  BATTERY   Thursday  01/15/2015   
2   140632022  SUSPICIOUS OCC  INVESTIGATIVE DETENTION  Wednesday  07/30/2014   
3   150383259         ASSAULT                  BATTERY   Saturday  05/02/2015   
4    40753980  OTHER OFFENSES         RECKLESS DRIVING     Friday  07/02/2004   

    Time PdDistrict      Resolution                   Address           X  \
0  01:00   SOUTHERN            NONE    800 Block of BRYANT ST -122.403405   
1  17:00    TARAVAL            NONE  1800 Block of VICENTE ST -122.485604   
2  09:32    BAYVIEW            NONE  100 Block of GILLETTE AV -122.396535   
3  23:10    BAYVIEW  ARREST, BOOKED   2400 Block of PHELPS ST -122.400131   
4  13:43    BAYVIEW            NONE   I-280 / CESAR CHAVEZ ST -120.500000   

           Y                                    Location            PdId  
0  37.775421   POINT (-122.403404791479 37.775420706711)  14619616171000  
1  37.738821   POINT (-122.48560378101 37.7388214326705)  15004567504134  
2  37.710660  POINT (-122.396535107224 37.7106603302503)  14063202264085  
3  37.730093  POINT (-122.400130573297 37.7300925390327)  15038325904134  
4  90.000000                           POINT (-120.5 90)   4075398065020

A column with the hour from the "Time" column is made. The Dataframe is also reduced to only obtain the focuscrimes. 

In [43]:
crime["Hour"] = pd.to_datetime(crime["Time"], format = "%H:%M").dt.hour # hour cloumn is made. 
focuscrimes = ['WEAPON LAWS', 'PROSTITUTION', 'DRIVING UNDER THE INFLUENCE', 'ROBBERY', 'BURGLARY', 'ASSAULT', 'DRUNKENNESS', 'DRUG/NARCOTIC', 'TRESPASS', 'LARCENY/THEFT', 'VANDALISM', 'VEHICLE THEFT', 'STOLEN PROPERTY', 'DISORDERLY CONDUCT']
crime = crime[crime["Category"].isin(focuscrimes)] #Reducing the dataframe only to obtain the forcuscrimes

The datafame is grouped by the hours coloumn and then counted by the amount of times the different categories are in the different hours intervals, this creates a new dataframe with the counts, of the different crimes, according to the hours of the day. Veiwing the first line, gives a good indication of what is happening. The index is reset. For some reason the index name cannot be deleted. **Hour 24 is hour 0 in our data**.

In [44]:
crimes_new = crime.groupby('Hour')['Category'].value_counts().unstack()
crimes_new.reset_index(level=0, inplace=True) # creating a new index, so hour becomes a column. 
crimes_new.head()

Category  Hour  ASSAULT  BURGLARY  DISORDERLY CONDUCT  \
0            0    10855      3496                 525   
1            1     9256      2267                 385   
2            2     8389      2443                 345   
3            3     4283      2517                 216   
4            4     2722      2233                 160   

Category  DRIVING UNDER THE INFLUENCE  DRUG/NARCOTIC  DRUNKENNESS  \
0                                 716           3775          830   
1                                 652           2303          785   
2                                 655           1858          664   
3                                 290           1260          259   
4                                 109            919          133   

Category  LARCENY/THEFT  PROSTITUTION  ROBBERY  STOLEN PROPERTY  TRESPASS  \
0                 19117          1377     2988              560       575   
1                 11789          1068     3064              394       402   
2                  7679           795     3196              330       453   
3                  4825           547     1923              256       370   
4                  3139           473     1246              228       275   

Category  VANDALISM  VEHICLE THEFT  WEAPON LAWS  
0              6534           4769         1154  
1              4562           3255          851  
2              4303           2577          759  
3              2898           1603          470  
4              1942           1238          317

Calculating the normalized arrests pr hour. In the code we are dividing every value with the corersponding sum of the respectivce column.  As we can see from the first lines, they are close the the small preview in the text. Therefor we can assume it is done right. 

In [45]:
for i in focuscrimes:
    total = crimes_new[i].sum()
    crimes_new[i] = crimes_new[i].apply(lambda x: x/total)
crimes_new.head()# Veiwing that it is done right

Category  Hour   ASSAULT  BURGLARY  DISORDERLY CONDUCT  \
0            0  0.055754  0.038190            0.052291   
1            1  0.047541  0.024764            0.038347   
2            2  0.043088  0.026687            0.034363   
3            3  0.021999  0.027495            0.021514   
4            4  0.013981  0.024393            0.015936   

Category  DRIVING UNDER THE INFLUENCE  DRUG/NARCOTIC  DRUNKENNESS  \
0                            0.126234       0.031556     0.084470   
1                            0.114951       0.019251     0.079890   
2                            0.115480       0.015531     0.067576   
3                            0.051128       0.010533     0.026359   
4                            0.019217       0.007682     0.013536   

Category  LARCENY/THEFT  PROSTITUTION   ROBBERY  STOLEN PROPERTY  TRESPASS  \
0              0.039790      0.082450  0.053484         0.047094  0.029565   
1              0.024538      0.063948  0.054845         0.033134  0.020669   
2              0.015983      0.047602  0.057207         0.027752  0.023292   
3              0.010043      0.032753  0.034421         0.021529  0.019024   
4              0.006533      0.028322  0.022303         0.019174  0.014140   

Category  VANDALISM  VEHICLE THEFT  WEAPON LAWS  
0          0.056299       0.037669     0.051902  
1          0.039308       0.025710     0.038275  
2          0.037076       0.020355     0.034137  
3          0.024970       0.012662     0.021139  
4          0.016733       0.009779     0.014257

#### 2. Creating the Bokeh plot.  
First a Bokeh dataframe is created from the calculated dataframe above 

In [46]:
source = ColumnDataSource(crimes_new)

The hours are converted from the dataframe are converted to a list,  and the datatype is a string, since the FactorRange operates with this datatype.

In [47]:
hours = [] # empty list to store the string converted hours
for i in range(24):
    hours.append(str(i))

The figure is created. Here the list created above is used to define the range of the x axis. The size of the plot is defined, and the title is added here.

In [48]:
p = figure(x_range = FactorRange(factors=hours),plot_height=500, plot_width= 1200, toolbar_location=None, title="Crimes")

The dictionary "bar" and the list "items" are created to store the information, derived from the following for loop, and then used to plot the information. Pallette Category20 has been used for the color scheme of the different categories. The muted_color has the same colors as the filled color. Others When untoggled, it will show a blue color, which can be a litte frustrating to look at.  

In [49]:
bar ={}
items = [] 
for indx,i in enumerate(focuscrimes):
    bar[i] = p.vbar(x="Hour", width=0.5,  top=i, source = source, muted_alpha = 0.06, alpha=3, muted_color=Category20[14][indx] ,color=Category20[14][indx]) 
    items.append((i, [bar[i]])) #the dictionary is appended to the items list, to the respective category.  


Additional information regarding the plot is added. This is the Axis titles, pushing the bars to the right, so they do not collide with the axis, but stand freely. 

In [50]:
legend = Legend(items=items, location=(0,50)) #Legend and location is configured
p.add_layout(legend, "left") #Adding the legend outisde the plot. 
p.xaxis.axis_label = "Hours of the day" #Setting x-axis title
p.yaxis.axis_label = "Normalized crimes pr hour" #Setting y-axis title
p.x_range.range_padding = 0.05 # 
p.legend.click_policy="hide" #mute is chossen, becuase this enables one to see all of the different categories partily if not chosen, instead of hiding them. 

show(p) #shows the plot

From the plot above we can see how the different crimes distribute according to each other in a neat way. 